### PR TITLE
Fix some issues in batchNormalization gradient

### PR DESCRIPTION
**Pull request by caisq**
Two issues fixed:
1. The outgoing gradient (`dy`) was previously not used.
2. The ordering of reduce sum and multiplication was wrong in
   some of the gradients previously.

Also tested with BatchNormalization training in tfjs-layers:
https://github.com/tensorflow/tfjs-layers/pull/139/files

### CODE REVIEW

**Comment by dsmilkov** [comment]
:lgtm_strong: This is great ! 

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[src/ops/batchnorm.ts, line 221 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZLGd5BoXyryJvbZly:-LAZLGd5BoXyryJvbZlz:b-8jy0k) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm.ts#L221)):*
> ```XML
>     const der = (dy: Tensor) => {
>       const scaleValue = scale == null ? ArrayOps.scalar(1) : scale;
>       const reductionAxes: number[] = [];
> ```

we have a util for this that might help: `broadcast_util.getReductionAxes(mean.shape, x4D.shape);`

First arg is inputShape, second is outputShape. Then it returns the reduction axes in order to bring outputSpace to inputSpace. See https://github.com/tensorflow/tfjs-core/blob/master/src/ops/binary_ops.ts#L64 for an example.

---

*[src/ops/batchnorm.ts, line 229 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZKwp0DkFzK4q7hrv3:-LAZKwp0DkFzK4q7hrv4:bh3jrvn) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm.ts#L229)):*
> ```XML
>         }
>         tileShape.push(1);
>       } else {
> ```

remove empty else

---

*[src/ops/batchnorm.ts, line 246 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZL96a9YPSG8fXTE69:-LAZL96a9YPSG8fXTE6A:b-v0j4s5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm.ts#L246)):*
> ```XML
>               .mul(scaleValue)
>               .reshape(x.shape);
>         } else {
> ```

no need for else { ... } since you return inside the if {...}

---

*[src/ops/batchnorm_test.ts, line 129 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZK45p6f_7LONj89_5:-LAZK45p6f_7LONj89_6:b4s6rh8) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm_test.ts#L129)):*
> ```XML
> 
>     const varianceEpsilon = .001;
> ```

not a blocker: are there useful tests in tfjs-layers that might be good to repeat in core as well, so other backends can benefit?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-:-LAZM0nBBHbg-XyFSrjZ:b-xbi9bb)*

**Comment by caisq** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/ops/batchnorm.ts, line 221 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZLGd5BoXyryJvbZly:-LAZPMDk93OxMG0T-LGI:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm.ts#L221)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

we have a util for this that might help: `broadcast_util.getReductionAxes(mean.shape, x4D.shape);`

First arg is inputShape, second is outputShape. Then it returns the reduction axes in order to bring outputSpace to inputSpace. See https://github.com/tensorflow/tfjs-core/blob/master/src/ops/binary_ops.ts#L64 for an example.
</blockquote></details>

Done.

---

*[src/ops/batchnorm.ts, line 229 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZKwp0DkFzK4q7hrv3:-LAZO0iE0gFRdjGhNd2f:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm.ts#L229)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

remove empty else
</blockquote></details>

Done.

---

*[src/ops/batchnorm.ts, line 246 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZL96a9YPSG8fXTE69:-LAZO4OB6S7STMImWULw:bkfas57) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm.ts#L246)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

no need for else { ... } since you return inside the if {...}
</blockquote></details>

I personally prefer this symmetrical return statements. Let me know if you feel strongly about it.

---

*[src/ops/batchnorm_test.ts, line 129 at r1](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970#-LAZK45p6f_7LONj89_5:-LAZOEjk0PurQ6Ohydvi:b-p8e8qb) ([raw file](https://github.com/tensorflow/tfjs-core/blob/c9dac6072ce2ddfbb28f386c9f69cf1d98ddf9a9/src/ops/batchnorm_test.ts#L129)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

not a blocker: are there useful tests in tfjs-layers that might be good to repeat in core as well, so other backends can benefit?
</blockquote></details>

Ack. But beware that adding such tests to core will involve some non-trivial work, because BatchNorm is a layer that keeps quite a few states: gamma, beta, step count, moving mean and moving variance.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/970)*
